cql3/restrictions/statement_restrictions.cc

```diff
@@ -1068,6 +1068,110 @@ std::vector<query::clustering_range> get_single_column_clustering_bounds(
     return ck_ranges;
 }
 
+// In old v1 indexes the token column was of type blob.
+// This causes problems because blobs are sorted differently than the bigint token values that they represent.
+// Tokens are encoded as 8 byte big endian two's complement signed integers,
+// which with blob sorting makes them ordered like this:
+// 0, 1, 2, 3, 4, 5, ..., bigint_max, bigint_min, ...., -5, -4, -3, -2, -1
+// Because of this clustering restrictions like token(p) > -4 and token(p) < 4 need to be translated
+// to two clustering ranges on the old index.
+std::vector<query::clustering_range> get_index_v1_token_range_clustering_bounds(
+        const query_options& options,
+        const column_definition* token_column,
+        const expression& token_restriction) {
+
+    // A workaround in order to make possible_lhs_values work properly.
+    // possible_lhs_values looks at the column type and uses this type's comparator.
+    // This is a problem because when using blob's comparator, -4 is greater than 4.
+    // This makes possible_lhs_values think that an expression like token(p) > -4 and token(p) < 4
+    // is impossible to fulfill.
+    // Create a fake token column with the type set to bigint, translate the restriction to use this column
+    // and use this restriction to calculate possible lhs values.
+    column_definition token_column_bigint = *token_column;
+    token_column_bigint.type = long_type;
+    expression new_token_restrictions = replace_column_def(token_restriction, &token_column_bigint);
+
+    std::variant<value_list, nonwrapping_range<managed_bytes>> values =
+        possible_lhs_values(&token_column_bigint, new_token_restrictions, options);
+
+    return std::visit(overloaded_functor {
+        [](const value_list& list) {
+            if (list.empty()) { // Impossible condition -- no rows can possibly match.
+                return std::vector<query::clustering_range>();
+            }
+
+            std::vector<query::clustering_range> ck_ranges;
+            ck_ranges.reserve(list.size());
+
+            for (auto&& value : list) {
+                ck_ranges.emplace_back(query::clustering_range::make_singular(std::vector<managed_bytes>{value}));
+            }
+
+            return ck_ranges;
```

Comment:
This looks like a replica of the get_single_column_clustering_bounds code at
https://github.com/scylladb/scylla/blob/7261c2c93e633a5df2772cc1313da59712ba74d4/cql3/restrictions/statement_restrictions.cc#L1066-L1068
But it's missing the sorting.  Will that be a problem?  Ideally, we could reuse the existing code here somehow.

Reply:
The code pushes first `[0, high)` and then `(low, -1]`. Thanks to this `ck_ranges` is already sorted. I checked manually and `sort` didn't change anything. Additionally I tried pushing the ranges in the opposite order and `sort` flipped it back to the current version.

Reply:
Are you saying this function only works correctly if the caller constructed token_restriction in a particular order?  That's... not ideal.  And again, we should reuse code: less typing, less chance of accidental bugs.

Reply:
I don't remember it perfectly, but I remember that last year I wrote a similar code. The ordering mattered, but in an unexpected way - when `ck_ranges` was in the opposite order I got wrong results, but only if the cache was empty (so the results didn't come back from cache, but directly from SSTables). I remember testing it by creating a table with old indexes, inserting some data, restarting Scylla and doing a select that would trigger this code path. The query would return wrong results, but it would start returning correct results after I did a full scan (and presumably filled the cache), suggesting that sstable reader expected sorted `ck_ranges`, but cache didn't.

Reply:
> Are you saying this function only works correctly if the caller constructed token_restriction in a particular order?

No, the order in which ranges are emplaced onto `ck_ranges` is indepdentent from how `token_restriction` is built.
These constraints are always fulfilled when `ck_ranges.size() > 1`:
* `token_low <= token_high`
* `token_low < 0`
* `token_high >= 0`

And then there is:
```c++
ck_ranges.emplace_back(std::move(zero_bound), std::move(upper_bound));
ck_ranges.emplace_back(std::move(lower_bound), std::move(min1_bound));
```

Which makes `ck_ranges` already sorted.

I guess I could add `sort()` just to be sure, but that is most likely not needed.

All `range_less` does in this case is
```c++
return cmp(x.start()->value(), y.start()->value());
```
And we know that `blob(zero) < blob(lower_bound)`

Reply:
Oh, I understand now.  It's because ck_ranges contents comes from a single value_list, which is
https://github.com/scylladb/scylla/blob/e4eb7df1a12d85a0afc6c367bc7246968fc86f7e/cql3/expr/expression.hh#L145
This code is fine as-is, then.